File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> currentPermutation;
        backtrack(nums, currentPermutation, ans);
        return ans;
    }

private:
    void backtrack(vector<int>& nums, vector<int>& currentPermutation, vector<vector<int>>& ans) {
        if (currentPermutation.size() == nums.size()) {
            ans.push_back(currentPermutation);
            return;
        }
        
        for (int i = 0; i < nums.size(); ++i) {
            if (find(currentPermutation.begin(), currentPermutation.end(), nums[i]) != currentPermutation.end()) {
                continue;
            }
            currentPermutation.push_back(nums[i]);
            backtrack(nums, currentPermutation, ans);
            currentPermutation.pop_back();
        }
    }
};

File: 0046-permutations/0046-permutations.py
from typing import List

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        ans = []
        stack = [(0, [False] * n, [])]

        while stack:
            i, vis, t = stack.pop()

            if i == n:
                ans.append(t)
                continue

            for j in range(n):
                if not vis[j]:
                   
                    new_vis = vis.copy()
                    new_vis[j] = True
                    stack.append((i + 1, new_vis, t + [nums[j]]))

        return ans